<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [ ]:
# текст запроса
query_3_1 = f'''
select COUNT(*) qty from public.vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

<ipython-input-28-7f364a65e0f6>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


qty
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [ ]:
# текст запроса
query_3_2 = f'''
select COUNT(*) qty from public.employers
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

<ipython-input-31-e0e2d23c59c3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


qty
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# текст запроса
query_3_3 = f'''
select COUNT(*) qty from public.areas
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

<ipython-input-33-11c9948ba1f3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


qty
0  1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# текст запроса
query_3_4 = f'''
select COUNT(*) qty from public.industries
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

# connection.close()

<ipython-input-9-ceb8d7fa2128>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


qty
0  294

*** Можно сделать предварительный вывод о том, что в среднем: \
- каждый работодатель разместил 2 вакансии; \
- в каждом регоне имеется около 36 вакансий от 17 работодателей; \
- на каждую сферу деятельности приходится 167 вакансий.

In [ ]:
# выводы по предварительному анализу данных

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''
select
  a.name,
  COUNT(v.id) cnt
from public.vacancies v
JOIN public.areas a ON v.area_id = a.id
GROUP BY a.id
ORDER BY cnt DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df
#df.head(60)

<ipython-input-33-73a25f114991>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0                         Москва  5333
1                Санкт-Петербург  2851
2                          Минск  2112
3                    Новосибирск  2006
4                         Алматы  1892
5                   Екатеринбург  1698
6                Нижний Новгород  1670
7                         Казань  1415
8                      Краснодар  1301
9                         Самара  1144
10                Ростов-на-Дону  1131
11                       Воронеж  1063
12                       Ташкент   851
13                    Красноярск   847
14                    Нур-Султан   840
15                     Челябинск   786
16                         Пермь   771
17                           Уфа   767
18                         Томск   692
19                          Омск   617
20                       Саратов   579
21                        Тюмень   547
22                   Владивосток   539
23                       Тбилиси   474
24                        Ижевск   466
25                     Волгоград   456
26                       Иркутск   451
27                     Ярославль   391
28                       Барнаул   374
29                        Рязань   346
30                   Калининград   343
31                          Тула   343
32                     Хабаровск   338
33                     Ульяновск   332
34                      Кемерово   316
35                         Пенза   305
36     Киров (Кировская область)   293
37                         Тверь   274
38                      Белгород   253
39                     Чебоксары   245
40                          Сочи   230
41                      Тольятти   229
42                       Армения   223
43                      Владимир   219
44                        Липецк   203
45              Набережные Челны   202
46                      Оренбург   199
47  Иваново (Ивановская область)   190
48                   Симферополь   187
49                         Курск   178
50                        Бишкек   166
51                        Брянск   161
52                        Калуга   158
53                   Новокузнецк   156
54                    Ставрополь   155
55                       Вологда   153
56                   Севастополь   152
57                      Кострома   146
58                      Смоленск   141
59                          Кипр   136

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''
select
  COUNT(id) cnt
from public.vacancies
WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

<ipython-input-16-bebb05645b92>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# текст запроса
query_4_3 = f'''
select
  ROUND(AVG(salary_from),0) mn_avg,
  ROUND(AVG(salary_to),0) mx_avg
from public.vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

<ipython-input-28-3cb76ea8c081>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


mn_avg    mx_avg
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''
select
  schedule,
  employment,
  COUNT(id) cnt
from public.vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

<ipython-input-30-ddd5034b1d4e>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''
select
  experience,
  COUNT(id) cnt
from public.vacancies
GROUP BY experience
ORDER BY cnt
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

# connection.close()

<ipython-input-32-868d7ae2deeb>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

*** Детальный анализ вакансий позволяет уточнить предварительный вывод. \
1) База данных содержит вакансии не только российских регионов. \
Вакансии распределены по регионам неравномерно: от нескольких тысяч в столицах и крупных городах до единичных предложений в населённых пунктах. \
2) Исходя из предположения, что указание заработной платы в вакансии является наиважнейшим критерием для отклика, половина вакансий в данной БД с пустыми полями для зарплаты может остаться без внимания соискателей. \
3) Диапазон средней заработной платы довольно высок, но он не точный, т.к. в половине вакансий зарплата не указана, и они выпадают из расчёта. \
4) Значительную долю (15,9%) после традиционной полной занятости "в офисе/на предприятии" (71,9%) занимают вакансии с типом графика "Удалённая работа". \
5) Наличие большой доли вакансий без опыта или с опытом до 3 лет (67,8%) с одной стороны может подтверждать гипотезу, что больше всего востребованы специалисты начального уровня, с другой стороны при современном темпе смены технологий больше ценится способность к переобучению, соответствие ценностям организации, чем годы опыта.

In [ ]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''
(select
  e.name,
  COUNT(e.id) cnt
from public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.id
ORDER BY cnt DESC
LIMIT 1)
UNION ALL
(select
  e.name,
  COUNT(e.id) cnt
from public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.id
ORDER BY cnt DESC
OFFSET 4 LIMIT 1)
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

<ipython-input-7-ad08e28dab7b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''
select
  a.name,
  COUNT(v.id) v_cnt,
  COUNT(e.id) e_cnt
from public.areas a
LEFT JOIN public.vacancies v ON a.id = v.area_id
LEFT JOIN public.employers e ON a.id = e.area
GROUP BY a.id
HAVING COUNT(v.id) = 0
ORDER BY e_cnt DESC
LIMIT 1
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

<ipython-input-7-5d775e2f9142>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  v_cnt  e_cnt
0  Россия      0    410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''
select
  e.name,
  COUNT(DISTINCT a.id) a_cnt
from public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
JOIN public.areas a ON v.area_id = a.id
GROUP BY e.id
ORDER BY a_cnt DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

<ipython-input-14-31d3e779128f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  a_cnt
0                                      Яндекс    181
1                                  Ростелеком    152
2                                  Спецремонт    116
3                      Поляков Денис Иванович     88
4                                   ООО ЕФИН      71
...                                       ...    ...
14901                            НПП Авиатрон      1
14902              Центр дистанционных торгов      1
14903  Городские Телекоммуникационные Системы      1
14904                       Введенский, Отель      1
14905              Группа Компаний Автоимпорт      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = f'''
select
  COUNT(id) cnt
from public.employers
WHERE id NOT IN (
    select employer_id from public.employers_industries
)
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

<ipython-input-9-ef2fed5e214c>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = f'''
select
  e.name
from public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id
HAVING COUNT(DISTINCT ei.industry_id) = 4
ORDER BY e.name
OFFSET 2 LIMIT 1
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

<ipython-input-11-a41bab2be398>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query_5_6 = f'''
select
  COUNT(DISTINCT e.id) cnt
from public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

<ipython-input-13-8e05d7ddda4a>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [ ]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

mln_cities = ''
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')

tbl = page.find('table', class_='standard sortable')
data_rows = tbl.find_all('tr')

for r in data_rows[1:]:
    mln_cities += "'" + r.a['title'] + "',"
mln_cities = mln_cities[:-1]

In [ ]:
# текст запроса
# способ 1
query_5_7 = f'''
select
  a.name,
  COUNT(v.id) cnt
from public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс'
and a.name IN (
  %s
)
GROUP BY a.name
UNION ALL
select
  'Total',
  COUNT(v.id) cnt
from public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс'
and a.name IN (
    %s
)
ORDER BY cnt
''' % (mln_cities, mln_cities)


# способ 2 (просто копируем города в виде последовательности для оператора IN)
# query_5_7 = f'''
# select
#   a.name,
#   COUNT(v.id) cnt
# from public.vacancies v
# JOIN public.employers e ON v.employer_id = e.id
# JOIN public.areas a ON v.area_id = a.id
# WHERE e.name = 'Яндекс'
# and a.name IN (
#   'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'
# )
# GROUP BY a.name
# UNION ALL
# select
#   'Total',
#   COUNT(v.id) cnt
# from public.vacancies v
# JOIN public.employers e ON v.employer_id = e.id
# JOIN public.areas a ON v.area_id = a.id
# WHERE e.name = 'Яндекс'
# and a.name IN (
#     'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'
# )
# ORDER BY cnt
# '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

<ipython-input-6-df759ebf7dc8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***
1) Количество вакансий у лидера работодателей - Яндекса - в 5,8 раза превышает количество вакансий крупной компании из топ-5 работодателей (Газпром нефть). Такая разница объясняется с одной стороны наличием целого ряда постоянно совершенствующихся технологических сервисов в Яндексе и специфической отраслевой направленностью в Газпром нефти. С другой стороны слишком большое количество вакансий может сигнализировать о текучести персонала. \
2) Результаты запроса скорее указывают на отличающиеся справочники регионов в таблицах AREAS, EMPLOYERS и VACANCIES, чем на реальных работодателей без вакансий в своём регионе: в частности регион Россия присутствует в первых двух таблицах и соответственно представлен в вакансиях в виде более мелкого деления по городам. \
3) Если опираться на количество вакансий, то можно предположить, что работодатели с количеством вакансий от 44 представлены хотя бы в половине регионов РФ. \
4) До 35% вакансий могут выпасть из расширенного поиска, т.к. в них не указана сфера деятельности работодателя. \
5) Наличие у работодателя как минимум 4-х сфер деятельности может указывать на то, что работодатель - крупная компания. \
6) Почти у 1/4 работодателей (23,6%) с указанной сферой деятельности занимаются разработкой ПО. \
7) Учитывая разнообразие и востребованность технологических сервисов Яндекса, ожидаемо, что вакансии данного работодателя есть во всех городах-миллионниках РФ.

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
# текст запроса
query_6_1 = f'''
select
  COUNT(id) cnt
from public.vacancies
WHERE LOWER(name) LIKE '%data%' or LOWER(name) LIKE '%данн%'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

<ipython-input-5-631a6abf7434>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [ ]:
# текст запроса
query_6_2 = f'''
SELECT
  COUNT(id) cnt
FROM public.vacancies
WHERE
(name ILIKE ANY(ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
or (name LIKE '%ML%' and name NOT LIKE '%HTML%'))
and (name ILIKE '%junior%' or experience='Нет опыта' or employment='Стажировка')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

<ipython-input-5-d5ce39ec296d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''
SELECT
  COUNT(id) cnt
FROM public.vacancies
WHERE
(name ILIKE ANY(ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
or (name LIKE '%ML%' and name NOT LIKE '%HTML%'))
and key_skills ILIKE ANY(ARRAY['%sql%', '%postgres%'])
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

<ipython-input-7-93be484d31ff>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''
SELECT
  COUNT(id) cnt
FROM public.vacancies
WHERE
(name ILIKE ANY(ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
or (name LIKE '%ML%' and name NOT LIKE '%HTML%'))
and key_skills ILIKE '%Python%'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

<ipython-input-5-a429304bcb54>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = f'''
SELECT
  ROUND(AVG(array_length(string_to_array(key_skills, '\t'), 1)), 2) avg_cnt
FROM public.vacancies
WHERE
(name ILIKE ANY(ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
or (name LIKE '%ML%' and name NOT LIKE '%HTML%'))
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

<ipython-input-11-99a1fdc5a2fc>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_cnt
0     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [ ]:
# текст запроса
query_6_6 = f'''
select
  experience,
  ROUND(AVG(CASE
    WHEN salary_from IS NOT NULL and salary_to IS NOT NULL THEN (salary_from + salary_to)/2
    ELSE COALESCE(salary_from, salary_to)
  END),0) avg_sal
FROM public.vacancies
WHERE
(name ILIKE ANY(ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])
or (name LIKE '%ML%' and name NOT LIKE '%HTML%'))
and NOT (salary_from IS NULL and salary_to IS NULL)
GROUP BY experience
HAVING experience = 'От 3 до 6 лет'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

<ipython-input-9-c2f1a4577ae9>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience   avg_sal
0  От 3 до 6 лет  243115.0

***
Вполне логично, что обучающийся на курсе Data Science захочет оценить свои карьерные перспективы, используя имеющуюся базу вакансий HeadHunter. \
Если осуществлять поиск по наличию уже привычного термина данные (а точнее его форм 'data' или 'данн') в названии вакансии, то помимо дата-саентистов в выборку попадут многочисленные администраторы, разработчики, архитекторы БД, бизнес-аналитики и т.п. Использование относительно новых понятий '%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%', %ML%' сократит выборку в 3 раза, но даст более точное совпадение со специализацией. \
Добавление фильтров по полям с навыками и опытом позволяет ещё точнее сопоставить личные возможности (+ представления) и требования (+ предложения) потенциальных работодателей. Имеющаяся база вакансий содержит как предложения для начинающих дата-саентистов, так и для имеющих опыт от 1 года, причём заработная плата для опытных кандидатов в 1,9 раза превышает среднемесячную зарплату по г. Москве (https://77.rosstat.gov.ru/folder/188598). В среднем работодатель ожидает от кандидата наличие около 6-ти профессиональных навыков, в целом Python упоминается в навыках для DS в 1,7 раза чаще, чем SQL или Postgres.

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту
Данная база вакансий содержит предложения о работе по регионам РФ, а также по странам ближнего и дальнего зарубежья: соответственно чем крупнее город (столица или город-миллионник), тем больше в нём вакансий. \
Большинство вакансий (67,8%) не требует опыта или с требованием к опыту до 3 лет. \
"Слабым местом" данной базы вакансий является отсутствие информации о заработной плате в половине из них и отсутствие сферы деятельности у 35% работодателей. Вакансии с отсутствующей информацией могут отсеяться при фильтрации потенциальным соискателем. Также следует отметить некоторое расхождение в делении по регионам работодателей и вакансий. \
Топ-5 сфер деятельности работодателей, априори подразумевающих широкие возможности трудоустройства и потенциал для карьерного роста: \
<table>
<thead>
<tr>
  <th>name</th>
  <th>cnt</th>
</tr>
</thead>
<tbody>
<tr>
  <td>Разработка программного обеспечения</td>
  <td>3553</td>
</tr>
<tr>
  <td>Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг</td>
  <td>2993</td>
</tr>
<tr>
  <td>Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее)</td>
  <td>1675</td>
</tr>
<tr>
  <td>Маркетинговые, рекламные, BTL, дизайнерские, Event-, PR-агентства, организация выставок</td>
  <td>798</td>
</tr>
<tr>
  <td>Консалтинговые услуги</td>
  <td>662</td>
</tr>
</tbody>
</table>

Наблюдается спрос на разработчиков ПО, специалистов с творческими и аналитическими способностями, растёт зависимость от цифровых платформ и технологий. \
Наиболее востребованными навыками - топ-10 - являются следующие: \
<table>
<thead>
<tr>
  <th>skill</th>
  <th>cnt</th>
</tr>
</thead>
<tbody>
<tr>
  <td>SQL</td>
  <td>6977</td>
</tr>
<tr>
  <td>Git</td>
  <td>4731</td>
</tr>
<tr>
  <td>Английский язык</td>
  <td>4525</td>
</tr>
<tr>
  <td>Работа в команде</td>
  <td>3923</td>
</tr>
<tr>
  <td>Linux</td>
  <td>3747</td>
</tr>
<tr>
  <td>JavaScript</td>
  <td>3625</td>
</tr>
<tr>
  <td>Python</td>
  <td>3445</td>
</tr>
<tr>
  <td>Управление проектами</td>
  <td>2764</td>
</tr>
<tr>
  <td>1С программирование</td>
  <td>2647</td>
</tr>
<tr>
  <td>PostgreSQL</td>
  <td>2452</td>
</tr>
</tbody>
</table>

Портрет идеального (самого разыскиваемого) кандидата: это \
СИСТЕМНЫЙ АДМИНИСТРАТОР (5,8%) с диапазоном средней заработной платы 49 667-61 477 руб. для работы в Санкт-Петербурге, Москве или Екатеринбурге, обладающий навыками настройки ПК, ПО, сетевого оборудования и знаниями Linux \
или ПРОГРАММИСТ 1С (5,4%) с диапазоном средней заработной платы 96 137-135 845 руб.  для работы в Краснодаре или Екатеринбурге, соответственно знающий линейку программных продуктов компании «1С», \
на полный рабочий день с опытом от 1 года до 3 лет.

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования